### PR TITLE
Assignment5 - Alexa Darrington and Johanne McClenahan

### PR DESCRIPTION
**Pull request by JohanneJayde**
Alexa's Github - @alyxadarr
Johanne's Github - @JohanneJayde

Website URL: https://aestheticwordle.com/

| Acount Type | Email                   | Password    |
| ----------- | ----------------------- | ----------- |
| Admin       | Admin@Intellitect.com   | P@ssw0rd123 |
| Awesome     | Awesome@Intellitect.com | P@ssw0rd123 |

Awesome should be able to update common word flag, delete word, and add word
Admin should only be able to update the common word flag

### CODE REVIEW

**Comment by montalcy** [comment]
Ill review this :)

File: Wordle.Api/Wordle.Api/Identity/Roles.cs [hunk 1]
```diff
@@ -2,4 +2,5 @@
 public static class Roles
 {
     public const string Admin = "Admin";
```
Comment (montalcy) [inline Wordle.Api/Wordle.Api/Identity/Roles.cs:4]: Just to clarify, awesome would technically be the one with the role MasterOfTheUniverse?

File: wordle-web/animations/fireworks.scss [hunk 2]
```diff
@@ -0,0 +1,124 @@
+@use "sass:math";
```
Comment (montalcy) [inline wordle-web/animations/fireworks.scss:1]: Really cool and creative styles!

File: wordle-web/components/SignInDialog.vue [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+          v-if="currentPage == 1"
+        />
+        <v-text-field v-model="userName" @keyup.stop label="Username" />
+        <v-text-field
+          v-model="password"
+          @keyup.stop
+          label="Password"
+          :type="showPassword ? 'text' : 'password'"
+          :append-inner-icon="
+            showPassword ? 'mdi-eye-off-outline' : 'mdi-eye-outline'
```
Comment (montalcy) [inline wordle-web/components/SignInDialog.vue:34]: Nice addition we talked about in class

**Review by montalcy — COMMENTED** [review]
Create a page named "WordEditor" that allows words to be edited ✅
Support deleting a word ✅
Support adding a new word ✅
Editing words is not supported ✅
Duplicate words should not be allowed ✅ --- Would be cool if you added an error message so the user knows its a duplicate word :)
Support changing the common word boolean flag ✅
Words should be sorted alphabetically ✅
Provide a text box for searching words AS YOU TYPE which supports searching for words starting with letters specified ✅
Provide a menu item for navigating to the WordEditor page ✅
Anyone can look at the word list ✅
Word list is paginated (10-100 per page) ✅
Any logged in user can change the common word flag ✅
There is a login page/component/popup/etc for users to sign in ✅
Only users over 21 years of age (based on birthday) with a claim of MasterOfTheUniverse can add and remove words ✅   
Users can be set up at start up in the API project as a seed (apriori), there does not need to be a sign up mechanism ✅


editing bc I didnt see the last bullet -- y'all did great. I've loved seeing the progress on this wordle!!

**Comment by connerhanson** [comment]
Ill review this as well!

File: Wordle.Api/Wordle.Api/Controllers/WordController.cs [hunk 4]
```diff
@@ -24,4 +27,38 @@ public async Task<string> GetWordOfDay(double offsetInHours = -7.0)
...
+    [Authorize(Policy = Policies.EditWord)]
+    public async Task DeleteWord(string word)
+    {
+        await wordEditorService.DeleteWord(word);
+    }
+
+    [HttpPost("EditWord")]
+    [Authorize]
+    public async Task EditWord(WordDto word)
+    {
```
Comment (connerhanson) [inline Wordle.Api/Wordle.Api/Controllers/WordController.cs:61]: Controllers all look great for the word Editor!

File: Wordle.Api/Wordle.Api/Identity/IdentitySeed.cs [hunk 5]
```diff
@@ -43,5 +50,24 @@ private static async Task SeedAdminUserAsync(UserManager<AppUser> userManager)
                 await userManager.AddToRoleAsync(user, Roles.Admin);
             }
         }
+
+        if (await userManager.FindByEmailAsync("Awesome@intellitect.com") == null)
+        {
```
Comment (connerhanson) [inline Wordle.Api/Wordle.Api/Identity/IdentitySeed.cs:55]: This looks great, is a little funny that the admin account has less permissions than the other, but functionally wise it all looks great.

File: Wordle.Api/Wordle.Api/Identity/Policies.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    public static void EditWordPolicy(AuthorizationPolicyBuilder policy)
+    {
+        policy.RequireClaim(Claims.MasterOfTheUniverse, "true");
+        policy.RequireAssertion(context =>
+        {
+            var birthdayString = context.User.Claims.FirstOrDefault(f => f.Type == Claims.BirthDate);
+            if (birthdayString != null)
+            {
+                DateTime birthday;
+                if (DateTime.TryParse(birthdayString.Value, out birthday))
```
Comment (connerhanson) [inline Wordle.Api/Wordle.Api/Identity/Policies.cs]: This works well, the only thing is that someone could be younger than 21 if their birthday hasn't happened yet that year.  You could fix this by doing something like I did in my attempt, Ill what I tried below: 

 int age = DateTime.Today.Year - paresedBDay.Year;
              // if this years date has not happened subtract a year from age
              if(DateTime.Today < paresedBDay.AddYears(age)){
                age--;
              }
              // else return if age is greater then 21
              return age >= 21;

Comment (JohanneJayde) [reply Wordle.Api/Wordle.Api/Identity/Policies.cs]: Yes I agree. I didn't even think about that until I did a review on Josh's and saw the checks he did. I will tri it out.

File: Wordle.Api/Wordle.Api/Services/WordEditorService.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        if(word is not null)
+        {
+            word.IsCommonWord = wordToEdit.IsCommonWord;
+
+            await Db.SaveChangesAsync();
+        }
+
+    }
+
+    public async Task AddWord(WordDto wordToAdd)
```
Comment (connerhanson) [inline Wordle.Api/Wordle.Api/Services/WordEditorService.cs:36]: Word editor services all look great!

File: Wordle.Api/Wordle.Api/Identity/IdentitySeed.cs (hunk 5 above)
Comment (JohanneJayde) [reply Wordle.Api/Wordle.Api/Identity/IdentitySeed.cs:55]: Yah, I'm probably going to update this maybe if we decide to keep working on it past the end of this class.

File: Wordle.Api/Wordle.Api/Controllers/WordController.cs (hunk 4 above)
Comment (JohanneJayde) [reply Wordle.Api/Wordle.Api/Controllers/WordController.cs:61]: Thank you 😄

File: wordle-web/components/SignInDialog.vue (hunk 3 above)
Comment (JohanneJayde) [reply wordle-web/components/SignInDialog.vue:34]: Thank you! Yah it was cool when they did in class. It's pretty simple logic

File: wordle-web/animations/fireworks.scss (hunk 2 above)
Comment (JohanneJayde) [reply wordle-web/animations/fireworks.scss:1]: Thank you! @alyxadarr did that for Assignment3 😄

File: Wordle.Api/Wordle.Api/Identity/Roles.cs (hunk 1 above)
Comment (JohanneJayde) [reply Wordle.Api/Wordle.Api/Identity/Roles.cs:4]: Yes. I added it in a weird way so I'm going to update the way we add the claim so its easier to see.

File: wordle-web/components/EditWordDialog.vue [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+watch(
+  () => props.word,
+  () => {
+    wordModel.value = props.isAdd ? "" : props.word?.word!;
+    isCommon.value = props.word?.isCommon ? "common" : "uncommon";
+  }
+);
+
+const validLengthWod = (value: string) => {
```
Comment (connerhanson) [inline wordle-web/components/EditWordDialog.vue:95]: The functionality seems to all work great from what I tested.

File: Wordle.Api/Wordle.Api/Services/WordEditorService.cs (hunk 7 above)
Comment (JohanneJayde) [reply Wordle.Api/Wordle.Api/Services/WordEditorService.cs:36]: Thank you 
a3f75e84b9542727230fc237ccc4e8d7

File: wordle-web/components/EditWordDialog.vue (hunk 8 above)
Comment (JohanneJayde) [reply wordle-web/components/EditWordDialog.vue:95]: 200w

Wahooooo

**Review by connerhanson — COMMENTED** [review]
Overall this site looks great, its been exciting watching it evolve.  The only thing I could recommend fixing is for the website not to redirect you back to the home page/reset your theme when you login.  Otherwise everything looks great!

### Create an editor for the available words ✅
  - Create a page named "WordEditor" that allows words to be edited ✅
    - Support deleting a word ✅
    - Support adding a new word ✅
    - Editing words is not supported ✅
    - Duplicate words should not be allowed ✅
    - Support changing the common word boolean flag ✅ 
    - Words should be sorted alphabetically ✅
  - Provide a text box for searching words AS YOU TYPE which supports searching for words starting with letters specified ✅
  - Provide a menu item for navigating to the WordEditor page ✅
    - Anyone can look at the word list ✅
    - Word list is paginated (10-100 per page) ✅
  - Any logged in user can change the common word flag ✅
    - There is a login page/component/popup/etc for users to sign in ✅
  - Only users over 21 years of age (based on birthday) with a claim of MasterOfTheUniverse can add and remove words ✅
  - Users can be set up at start up in the API project as a seed (apriori), there does not need to be a sign up mechanism ✅

**Comment by MikaellaCroskrey-IntelliTect** [comment]
## Assignment 5


### Create an editor for the available words ✅
  - Create a page named "WordEditor" that allows words to be edited ✅
    - Support deleting a word ✅
    - Support adding a new word ✅
    - Editing words is not supported ✅
    - Duplicate words should not be allowed ✅
    - Add & support changing the common word boolean flag ✅
      - Filter the list of returned words by common words ❌ Didn't see a way to filter the list in the UI or implemented code 
    - Words should be sorted alphabetically ✅
  - Provide a text box for searching words AS YOU TYPE which supports searching for words starting with letters specified ✅
  - Provide a menu item for navigating to the WordEditor page ✅
    - Anyone can look at the word list ✅
    - Word list is paginated on the server side (10-100 per page) ✅
  - Any logged in user can change the common word flag ✅
    - There is a login page/component/popup/etc for users to sign in ✅
  - Only users over 21 years of age (based on birthday) with a claim of MasterOfTheUniverse can add and remove words ✅ Great job on the claims!
  - Users can be set up at start up in the API project as a seed (apriori), there does not need to be a sign up mechanism ✅


## Extra Credit
Find a bug in the application, create an issue, submit a pull request against the issue ❌
Add the ability to sign up for an account by inputting an email and password ❌ Attempted but not finished?
Allow the user to elevate their level of privelage by inputting a secret phrase ❌
Select pagination record count ✅

Awesome work! The code looked really good, there was great attention to detail, and it had a very aesthetic UI. :)

**Pull request closed without merging**